A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,398 @@
+    String encoded_address = base64_encode(hostAddress);
+    Create a Cookie from our cookie configuration;
+    Set cookie's value as encoded_address;
+    Add the cookie header to the response;
+  }
+```
+
+### LB Policy for Stateful Session Affinity
+
+After the `CookieBasedStatefulSessionFilter` has passed the `override-host`
+value to the Load Balancer as a "pick-argument", the Load Balancer uses
+this value to route the RPC appropriately. The required logic needs to be
+implemented in a new and separate LB policy at an appropriate place in the
+hierarchy.
+
+RPC load reporting happens in the `“xds_cluster_impl”` policy and we do want
+all RPCs to be included in the load reports. And this policy conditionally
+routes an RPC based on the `override-host` value if present, or else delegates
+to the currently configured LB policy. Hence the new policy needs to be just
+below the `“xds_cluster_impl”` as its child policy. Let's call this new policy
+`“override-host-experimental”`. This policy contains subchannel management,
+endpoint management, and RPC routing logic as follows:
+
+* maintain a map of valid endpoints (or `ResolvedAddresses`) and keep it
+  up-to-date based on endpoint updates from the resolver. Let's call this
+  map `validEndpoints` where the key is an address (IP:port) and the value
+  is a list of equivalent-addresses (from the endpoint updates).
+
+* maintain a map - let’s call this `overrideHostMap`. The key is (IP:port)
+  and the value is the tuple
+  `[subchannel, list of other equivalent addresses]`. When a host (endpoint)
+  has multiple addresses (e.g. IPv6 vs IPv4 or due to multiple network
+  interfaces) they are said to be equivalent addresses.
+
+* whenever a new subchannel is created (by the child policy that is
+  routing an RPC - see below), add a new entry to `overrideHostMap` with
+  the subchannel address (i.e. the peer address) as the key and the value as
+  `[subchannel, other equivalent addresses of the subchannel]`. The "other
+  equivalent addresses of the subchannel" are a property of the subchannel.
+
+    * we may have to wait for subchannel to be `READY` in order to know which
+      specific address the subchannel is connected to before we add it to the
+      map.
+
+    * in C-core and Node, LB policies create a new subchannel for every address
+      every time there is a new address list. In the map, we will just replace
+      an existing entry with the most recent subchannel created for a given
+      address i.e. the new subchannel created will replace the older one for
+      that address.
+
+* whenever a subchannel is shut down, remove the associated entries from
+  `overrideHostMap`.
+
+    * this can be achieved by wrapping a subchannel to intercept subchannel
+      shutdown and the wrapper is returned by `createSubchannel`.
+
+* the policy's subchannel picker pseudo-code is as follows:
+
+```
+   func pickSubchannel() {
```

Comment:
One thing I noticed in this pseudo-code is you don't recursively try multiple levels of equivalent addresses. e.g. origin address from override_host is A and the entry for A has (B,C) as equivalent addresses.  B's entry also has equivalent addresses (F,G) and until we find a ready sub-channel we need to look at those as well. Don't we?

Reply:
fixed the pseudo-code based on the comment

Reply:
I don't think we want to do this recursive expansion of equivalent addresses in the picker, since that will add more overhead on a per-pick basis.  Instead, I think we want to do this expansion when we populate the map.

For example, let's say that we get an update from our parent with the following equivalent address groups:
- [A, B]
- [B, C]
- [D, E]
- [E, F]

In this case, here are the lists of equivalent addresses that we would construct for each address, which would be stored in the map value for each address:
- A: [B, C]
- B: [A, C]
- C: [A, B]
- D: [E, F]
- E: [D, F]
- F: [D, E]

I think the picker pseudo-code I provided above is correct as-is.  Let's use that in the doc.

Also, please provide pseudo-code for how we compute the equivalent addresses for each address in the map.  We need to make sure all implementations will use the same algorithm here.

Reply:
Okay, your idea sounds better. In terms of pseudo-code for computing equivalent addresses, you are just suggesting codifying your example, right?